### PR TITLE
Change Net.Mail to Mailkit for Smtpclient.

### PR DESCRIPTION
**Pull request by kcherukupalli**
## What’s changing?

The SmtpClient class is deprecated and refactored to use Mailkit as per Microsoft recommendation. 
 
https://learn.microsoft.com/en-us/dotnet/api/system.net.mail.smtpclient?view=net-6.0#remarks

## Why?

Nac-1574

## This PR has:

- [x] **JIRA ticket(s)** (_Must be listed in PR request and branch name should follow convention NAC-123-description_)<description>

- [ ] **Security Related Changes** (_Security Tab questions have been answered on JIRA ticket(s)_)

- [x] **Commented source code** (_required on public classes and methods, and elsewhere as appropriate_)

- [ ] **Documented API changes** (_required on public classes and methods, and elsewhere as appropriate_)
    - [ ] External API documentation - OpenApi docs and Generated Docs (e.g. https://github.com/18F/piipan/tree/dev/docs/openapi) 
    - [ ] Internal API documentation (e.g. https://github.com/18F/piipan/tree/dev/metrics/docs/openapi)   

- [ ] **Developer documentation** (_for new build/test/API changes or complex portions of the system_)

- [ ] **Automated unit tests** (_to maintain or increase level of code coverage_)

- [ ] **Changes to IAC** (_add any deployment steps that require manual intervention to the draft release notes_)
    - [ ] Have IaC scripts run successfully against an existing deployed environment?
    - [ ] Have IaC scripts run successfully against a newly deployed environment?
    - [ ] Have any new bash scripts been given the appropriate permissions? (e.g. git update-index --chmod=+x $script)
    - [ ] Do the IaC changes require deleting existing resources?

## Anything else?

_Add other details that are helpful for review here_

## Code Review Checklist:

- [ ] Automated checks all pass? (code builds successfully, tests pass, test coverage meets/exceeds threshold, etc.) 
- [ ] Have all requirements and Acceptance Criteria been met?
- **If there are code changes**
    - [ ] Is there no unnecessary complexity? Code is readable, maintainable, etc.
    - [ ] Do the changes follow SOLID principles?
    - [ ] Are the changes robust? Concurrency considered? Performance? Security (e.g. no sql injections)?
    - [ ] Are newly added dependencies needed and appropriate? Are dependency files (e.g. packages.json & packages.lock.json) updated?
    - [ ] For any new projects/solutions, have they been added to code coverage files?
    - [ ] Public classes & methods are documented    
- **If there are API changes**
    - [ ] Does API make sense? Useful and concise?
    - [ ] Are existing API interfaces/contracts intact? i.e. no breaking changes
    - [ ] Are API tests included?
    - [ ] Are API responses appropriate and documented? (200s, 201, 404s, etc.)
    - [ ] Are internals kept private and not leaked into the API?
    - [ ] External API documentation - OpenApi docs and Generated Docs 
    - [ ] Internal API documentation    
- **If there are IAC changes**
    - [ ] IAC Scripts are readable & maintainable
    - [ ] IAC Scripts are documented
    - [ ] IAC scripts pass required checks
    - [ ] IAC scripts were run successfully by author
- **General Logging Questions**
    - [ ] Do new capabilities (or changed capabilities) include logging?
    - [ ] Is PII not logged?
- **General Error Handling Questions**
    - [ ] Has exception handling been added for new capabilities (or changed capabilities)?
    - [ ] Are errors/exceptions properly communicated to end-users? Were there requirements for user facing error messages?
    - [ ] Are errors/exceptions logged?
    - [ ] Is PII kept out of error/exception logging?
- **If there are Developer Documentation Changes**
    - [ ] General Project documentation (under piipan\docs. e.g. iac.md, architecture.md, etc.)
    - [ ] Github documentation additions/changes (Readmes)
- **General automated test questions**
    - [ ] End-to-End Tests provided (if code implements new user story)
    - [ ] Are tests sufficient?
    - [ ] Corner cases considered?
    - [ ] Are performance tests needed?

### CODE REVIEW

File: notifications/src/Piipan.Notifications.Core/Services/MailService.cs [hunk 1]
```diff
@@ -30,57 +34,62 @@ public async Task<bool> SendEmailAsync(EmailModel emailModel, ILogger logger, Ac
...
-                var res = _usdaImageRetriever.RetrieveUsdaSymbolColorImageResource();
-                res.ContentId = "usda-img";
-                AlternateView alternateView = AlternateView.CreateAlternateViewFromString(emailModel.Body, null, MediaTypeNames.Text.Html);
-                alternateView.LinkedResources.Add(res);
-                mailMessage.AlternateViews.Add(alternateView);
-                mailMessage.To.Add(string.Join(',', emailModel.ToList.Where(s => !string.IsNullOrWhiteSpace(s))));
+                var builder = new BodyBuilder();
+                var image = builder.LinkedResources.Add(_usdaImageRetriever.RetrieveUsdaSymbolColorImagePath());
+                image.ContentId = "usda-img";
+                builder.LinkedResources.Add(image);
```
Comment (bgfoshay) [inline notifications/src/Piipan.Notifications.Core/Services/MailService.cs:46]: Question: Is this adding the image to the LinkedResources a second time?

File: notifications/src/Piipan.Notifications.Core/Services/MailService.cs [hunk 2]
```diff
@@ -30,57 +34,62 @@ public async Task<bool> SendEmailAsync(EmailModel emailModel, ILogger logger, Ac
...
+                    }
                 }
 
                 if (_emailDelivery.Enabled)
                 {
-                    smtpClient.Send(mailMessage);
-                    logger.LogInformation($"Email sent successful. " + mailMessage.ToStringExtended(emailModel.Body));
+                    _smtpClient.Connect(Environment.GetEnvironmentVariable("SmtpServer"), 587, SecureSocketOptions.None);
+                    _smtpClient.Send(mimeMessage);
+                    _smtpClient.Disconnect(true);
```
Comment (bgfoshay) [inline notifications/src/Piipan.Notifications.Core/Services/MailService.cs:75]: Question: what happens if Disconnect doesn't get called? i.e. what happens if Send fails? Do we need a try/finally block?

File: notifications/tests/Piipan.Notifications.Core.Tests/Services/MailServiceTest.cs [hunk 3]
```diff
@@ -23,14 +24,10 @@ public async void FailedSendEmailTest()
...
-            retrieverSetup.Setup(x => x.RetrieveUsdaSymbolColorImageResource()).Returns(LinkedResource.CreateLinkedResourceFromString("abc"));
-
-            var smtpClient = new SmtpClient(Environment.GetEnvironmentVariable(SmtpConnection))
-            {
-                Port = 587,
-                EnableSsl = false,
-            };
+            retrieverSetup.Setup(x => x.RetrieveUsdaSymbolColorImagePath()).Returns("images/usda-symbol-color.png");
 
+            var smtpClient = new SmtpClient();
```
Comment (bgfoshay) [inline notifications/tests/Piipan.Notifications.Core.Tests/Services/MailServiceTest.cs:29]: Thought: I think we should verify that Send gets called (or not called when it shouldn't be called) for all of these tests.

File: notifications/src/Piipan.Notifications.Core/Services/MailService.cs [hunk 4]
```diff
@@ -30,57 +34,62 @@ public async Task<bool> SendEmailAsync(EmailModel emailModel, ILogger logger, Ac
         {
             try
             {
-                var mailMessage = new MailMessage()
-                {
-                    Subject = emailModel.Subject,
-                    From = new MailAddress(_emailDelivery.EmailFrom),
-                    IsBodyHtml = true
-                };
+                var mimeMessage = new MimeMessage();
```
Comment (bgfoshay) [inline notifications/src/Piipan.Notifications.Core/Services/MailService.cs:37]: Suggestion: It might be worth pulling all of the MimeMessage construction out into another interface/class e.g. IMessageBuilder.cs with public MimeMessage ConstructMimeMessage(...). Then we can mock out that class/method in the MailServiceTests.cs and add a separate set of tests that ensure the message is constructed correctly.

File: notifications/tests/Piipan.Notifications.Core.Tests/Services/MailServiceTest.cs [hunk 5]
```diff
@@ -332,5 +333,77 @@ public async void SuccessSendEmail_When_CC_IsEmpty_Test()
...
+            Assert.True(result);
+            Assert.False(errorEncountered);
+            logger.Verify(n => n.Log(
+                 It.Is<LogLevel>(l => l == LogLevel.Information),
+                 It.IsAny<EventId>(),
+                 It.Is<It.IsAnyType>((object v, Type _) => v.ToString().Contains($"Email sent successful. Email Subject:")),
+                 It.IsAny<Exception>(),
+                 (Func<It.IsAnyType, Exception, string>)It.IsAny<object>()),
+               Times.Once());
+            smtpClient.Verify(m => m.Send(It.Is<MimeMessage>(p=> p.GetType().Name.Equals(mimeMessage.GetType().Name)), default, null), Times.Once);
```
Comment (bgfoshay) [inline notifications/tests/Piipan.Notifications.Core.Tests/Services/MailServiceTest.cs:405]: Question: Does it work if you use ReferenceEquals() here instead of checking GetType().Name? https://learn.microsoft.com/en-us/dotnet/api/system.object.referenceequals?view=net-7.0. Something like Object.ReferenceEquals(p, mimeMessage)?

File: notifications/tests/Piipan.Notifications.Core.Tests/Services/MessageBuilderTest.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            
+            Assert.Equal(message.To.ToString(), mimeMessage.To.ToString());
+            Assert.Equal(message.Cc.ToString(), mimeMessage.Cc.ToString());
+            Assert.Equal(message.Bcc.ToString(), mimeMessage.Bcc.ToString());
+
+            Assert.Equal(message.Subject, mimeMessage.Subject);
+            Assert.True(mimeMessage.HtmlBody.Contains("Test email body."));
+
+            // Check the image and content id
+          //  Assert.True(mimeMessage.Body.ToString().Contains(contentId));
```
Comment (bgfoshay) [inline notifications/tests/Piipan.Notifications.Core.Tests/Services/MessageBuilderTest.cs:102]: Question/Nitpick: Are mimeMessage.Body.ContentId and mimeMessage.Body.ContentType.Name not correct?

File: notifications/tests/Piipan.Notifications.Core.Tests/Services/MailServiceTest.cs (hunk 5 above)
Comment (kcherukupalli) [reply notifications/tests/Piipan.Notifications.Core.Tests/Services/MailServiceTest.cs:405]: With ReferenceEquals test fails.

File: notifications/src/Piipan.Notifications.Func.Api/Startup.cs [hunk 7]
```diff
@@ -32,16 +32,14 @@ public override void Configure(IFunctionsHostBuilder builder)
...
-                return new SmtpClient(Environment.GetEnvironmentVariable(SmtpConnection))
-                {
-                    Port = 587,
-                    EnableSsl = false,
-                };
-            });
+            builder.Services.AddTransient<IMessageBuilder, MessageBuilder>();
             builder.Services.AddTransient<IMailService, MailService>();
             builder.Services.AddTransient<IUsdaImageRetriever, UsdaImageRetriever>();
+            builder.Services.AddTransient<ISmtpClient>(x =>
```
Comment (dborgen2) [inline notifications/src/Piipan.Notifications.Func.Api/Startup.cs:38]: Question: Can we just do `AddTransient<ISmtpClient, SmtpClient>()`?

File: notifications/tests/Piipan.Notifications.Core.Tests/Services/MessageBuilderTest.cs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+namespace Piipan.Notifications.Core.Tests.Services
+{
+    public class MessageBuilderTest
+    {
+
+        [Fact]
+        public async void MessageBuilder_Test()
+        {
+            // Arrange
+            Environment.SetEnvironmentVariable("SmtpCcEmail", "test-cc@example.com");
```
Comment (dborgen2) [inline notifications/tests/Piipan.Notifications.Core.Tests/Services/MessageBuilderTest.cs:20]: After each test can we clear the environment variables?

File: notifications/tests/Piipan.Notifications.Core.Tests/Services/MessageBuilderTest.cs [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+            Environment.SetEnvironmentVariable("SmtpBccEmail", "test-bcc@example.com");
+            Environment.SetEnvironmentVariable("SmtpFromEmail", "noreply@example.com");
+            var services = new ServiceCollection();
+            string SmtpConnection = "Smtpserver";
```
Comment (dborgen2) [inline notifications/tests/Piipan.Notifications.Core.Tests/Services/MessageBuilderTest.cs:24]: Typo here, should be SmtpServer

File: notifications/src/Piipan.Notifications.Core/Services/IMessageBuilder.cs [hunk 10]
```diff
@@ -0,0 +1,11 @@
+﻿using MimeKit;
+using Piipan.Notifications.Core.Extensions;
+using Piipan.Notifications.Models;
+
+namespace Piipan.Notifications.Core.Services
+{
+    public interface IMessageBuilder
```
Comment (dborgen2) [inline notifications/src/Piipan.Notifications.Core/Services/IMessageBuilder.cs:7]: Can you comment this new interface?

File: notifications/src/Piipan.Notifications.Core/Services/MessageBuilder.cs [hunk 11]
```diff
@@ -0,0 +1,58 @@
+﻿using Microsoft.Extensions.Logging;
+using MimeKit;
+using Piipan.Notifications.Core.Extensions;
+using Piipan.Notifications.Models;
+
+namespace Piipan.Notifications.Core.Services
+{
+    /// <summary>
+    /// Mimemessage builder
```
Comment (dborgen2) [inline notifications/src/Piipan.Notifications.Core/Services/MessageBuilder.cs:9]: Question: This seems like maybe a more detailed summary is needed?

File: notifications/tests/Piipan.Notifications.Core.Tests/Services/MailServiceTest.cs [hunk 12]
```diff
@@ -359,5 +341,196 @@ public async void SuccessSendEmail_When_CC_IsEmpty_Test()
                  (Func<It.IsAnyType, Exception, string>)It.IsAny<object>()),
                Times.Once());
         }
+
+        [Fact]
+        public async Task BuildMessageBody_Test()
```
Comment (dborgen2) [inline notifications/tests/Piipan.Notifications.Core.Tests/Services/MailServiceTest.cs:346]: Question: Can we make the name of this test a bit more descriptive. What exactly is it testing?

File: notifications/src/Piipan.Notifications.Core/Services/MessageBuilder.cs [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+        public MimeMessage ConstructMimeMessage(EmailModel emailModel, string footerImagePath, EmailDelivery emailDelivery)
+        {
+            var mimeMessage = new MimeMessage();
+            mimeMessage.Subject = emailModel.Subject;
+            var builder = new BodyBuilder();
+            var image = builder.LinkedResources.Add(footerImagePath);
+            image.ContentId = "usda-img";
+            builder.HtmlBody += emailModel.Body;
+            mimeMessage.Body = builder.ToMessageBody();
+            mimeMessage.From.Add(MailboxAddress.Parse(emailDelivery.EmailFrom));
```
Comment (dborgen2) [inline notifications/src/Piipan.Notifications.Core/Services/MessageBuilder.cs:31]: Question: Do we need to worry about invalid emails causing this Parse to fail?

File: notifications/src/Piipan.Notifications.Func.Api/Startup.cs (hunk 7 above)
Comment (kcherukupalli) [reply notifications/src/Piipan.Notifications.Func.Api/Startup.cs:38]: We are creating a singleton for smtpclient.

File: notifications/tests/Piipan.Notifications.Core.Tests/Services/MailServiceTest.cs [hunk 14]
```diff
@@ -20,7 +21,12 @@ public MailServiceTest()
         {
             _messageBuilder = new MessageBuilder();
         }
-
+        private void ClearEnvironmentVariables()
+        {
+            Environment.SetEnvironmentVariable("SmtpCcEmail", null);
+            Environment.SetEnvironmentVariable("SmtpBccEmail", null);
+            Environment.SetEnvironmentVariable("SmtpFromEmail", null);
```
Comment (dborgen2) [inline notifications/tests/Piipan.Notifications.Core.Tests/Services/MailServiceTest.cs:28]: Nitpick (non-blocking): I think we also need to clear the SmtpServer variable here.

**Review by dborgen2 — APPROVED** [review]
👍
